src/main/java/com/google/devtools/build/lib/remote/GrpcRemoteExecutorKeepalived.java

```diff
@@ -0,0 +1,320 @@
+      }
+    }
+
+    @Nullable
+    ExecuteResponse waitExecution() throws IOException {
+      Preconditions.checkState(lastOperation != null);
+
+      WaitExecutionRequest request = WaitExecutionRequest.newBuilder()
+          .setName(lastOperation.getName())
+          .build();
+      try {
+        Iterator<Operation> operationStream = executionBlockingStubSupplier.get()
+            .waitExecution(request);
+
+        // We want to reset backoff for WaitExecution() so we can "infinitely" wait for the
+        // execution to complete as long as they are making progress (by returning response).
+        return handleOperationStream(operationStream, /* resetWaitExecutionBackoff */ true);
+      } catch (StatusRuntimeException e) {
+        // A NOT_FOUND error means Operation was lost on the server, retry Execute().
+        //
+        // However, we only retry Execute() if executeBackoff should retry. Also increase the retry
+        // counter at the same time since we will technically retry (done by nextDelayMillis()).
+        if (e.getStatus().getCode() == Code.NOT_FOUND && executeBackoff.nextDelayMillis() >= 0) {
+          lastOperation = null;
+          return null;
+        }
+        throw new IOException(e);
+      }
+    }
+
+    /** Process a stream of operations from Execute() or WaitExecution(). */
+    ExecuteResponse handleOperationStream(Iterator<Operation> operationStream,
+        boolean resetWaitExecutionBackoff) throws IOException {
+      try {
+        while (operationStream.hasNext()) {
+          Operation operation = operationStream.next();
+          ExecuteResponse response = handleOperation(operation);
+
+          // At this point, we received the response without error.
+          lastOperation = operation;
+
+          if (resetWaitExecutionBackoff) {
+            waitExecutionBackoff.reset();
+          }
+
+          // Update execution progress to the observer.
+          //
+          // After called `execute` above, the action is actually waiting for an available gRPC
+          // connection to be sent. Once we get a reply from server, we know the connection is up
+          // and indicate to the caller the fact by forwarding the `operation`.
+          //
+          // The accurate execution status of the action relies on the server
+          // implementation:
+          //   1. Server can reply the accurate status in `operation.metadata.stage`;
+          //   2. Server may send a reply without metadata. In this case, we assume the action is
+          //      accepted by the server and will be executed ASAP;
+          //   3. Server may execute the action silently and send a reply once it is done.
+          observer.onNext(operation);
+
+          if (response != null) {
+            return response;
+          }
+        }
+
+        // The operation completed successfully but without a result.
+        throw new IOException("Remote server error: execution terminated with no result.");
```

Comment:
nit: remove dot from the end of the exception message (IIRC that's the convention around here)

Reply:
Done.